lib/puppet/provider/cisco_route_map/cisco.rb

```diff
@@ -0,0 +1,588 @@
+        actions.each do |action, nu_obj|
+          rm_instances << properties_get(rmname, sequence, action, nu_obj)
+        end
+      end
+    end
+    rm_instances
+  end # self.instances
+
+  def self.prefetch(resources)
+    rm_instances = instances
+    resources.keys.each do |id|
+      provider = rm_instances.find do |rmi|
+        rmi.rmname.to_s == resources[id][:rmname].to_s &&
+        rmi.sequence.to_s == resources[id][:sequence].to_s &&
+        rmi.action.to_s == resources[id][:action].to_s
+      end
+      resources[id].provider = provider unless provider.nil?
+    end
+  end
+
+  def exists?
+    @property_hash[:ensure] == :present
+  end
+
+  def create
+    @property_flush[:ensure] = :present
+  end
+
+  def destroy
+    @property_flush[:ensure] = :absent
+  end
+
+  def properties_set(new_rm=false)
+    ROUTE_MAP_ALL_PROPS.each do |prop|
+      next unless @resource[prop]
+      send("#{prop}=", @resource[prop]) if new_rm
+      unless @property_flush[prop].nil?
+        @nu.send("#{prop}=", @property_flush[prop]) if
+          @nu.respond_to?("#{prop}=")
+      end
+    end
+    # custom setters which require one-shot multi-param setters
+    match_community_set
+    match_ext_community_set
+    match_ipv4_multicast_set
+    match_ipv6_multicast_set
+    match_ip_addr_access_list
+    match_ip_addr_prefix_list
+    match_route_type_set
+    set_dampening_set
+    set_distance_set
+    set_community_set
+    set_extcommunity_4bytes_set
+    set_extcommunity_rt_set
+    set_extcommunity_cost_set
+    set_ip_next_hop_set
+    set_ip_precedence
+    set_metric_set
+  end
+
+  def match_ip_addr_access_list
+    pf = @property_flush[:match_ipv4_addr_access_list]
+    v4 = pf.nil? ? @nu.match_ipv4_addr_access_list : pf
+    pf = @property_flush[:match_ipv6_addr_access_list]
+    v6 = pf.nil? ? @nu.match_ipv6_addr_access_list : pf
+    @nu.match_ip_addr_access_list(v4, v6)
+  end
+
+  def match_ip_addr_prefix_list
+    pf = @property_flush[:match_ipv4_addr_prefix_list]
+    v4 = pf.nil? ? @nu.match_ipv4_addr_prefix_list : pf
+    pf = @property_flush[:match_ipv6_addr_prefix_list]
+    v6 = pf.nil? ? @nu.match_ipv6_addr_prefix_list : pf
+    @nu.match_ip_addr_prefix_list(v4, v6)
+  end
+
+  def match_community_set
+    comm = @property_flush[:match_community] ? @property_flush[:match_community] : @nu.match_community
+    pf = @property_flush[:match_community_exact_match]
+    exact = PuppetX::Cisco::Utils.flush_boolean?(pf) ? pf : @nu.match_community_exact_match
+    @nu.match_community_set(comm, exact)
+  end
+
+  def match_ext_community_set
+    comm = @property_flush[:match_ext_community] ? @property_flush[:match_ext_community] : @nu.match_ext_community
+    pf = @property_flush[:match_ext_community_exact_match]
+    exact = PuppetX::Cisco::Utils.flush_boolean?(pf) ? pf : @nu.match_ext_community_exact_match
+    @nu.match_ext_community_set(comm, exact)
+  end
+
+  def match_ipv4_multicast_set
```

Comment:
@saichint In general, we should make sure that properties that have dependencies on other properties are handles in the type.